cpp/spherical_volume_rendering_util.cpp

```diff
@@ -0,0 +1,185 @@
+#include "spherical_volume_rendering_util.h"
+#include <vector>
+#include <cmath>
+
+// Takes the absolute value of 'value' and determines if it is less than the tolerance.
+// In this case, we assume tolerance is an arbitrary small value.
+// This is used when floating point mathematics carries rounding error.
+[[nodiscard]] inline bool isNearZero(double value, double tolerance) {
+    return std::abs(value) < tolerance;
+}
+
+// Produces the corresponding angular or azimuthal voxel ID given two directions.
+// TODO(cgyurgyik): Update documentation, provide link to better description.
+[[nodiscard]] inline size_t getInitialVoxelID(double y, double x, size_t num_sections) {
+    const size_t current_voxel_ID = std::floor(std::atan2(y, x) * num_sections / ( 2 * M_PI));
+    return (current_voxel_ID + num_sections) % num_sections;
+}
+
+struct RadialHitParameters {
+    double tMaxR;
+    size_t tStepR;
+    bool previous_transition_flag;
+};
+
+struct AngularHitParameters {
+    double tMaxTheta;
+    size_t tStepTheta;
+};
+
+struct AzimuthalHitParameters {
+    double tMaxPhi;
+    size_t tStepPhi;
+};
+
+// TODO(cgyurgyik): Documentation, implementation.
+[[nodiscard]] RadialHitParameters radialHit(const Ray& ray, const SphericalVoxelGrid& grid, size_t current_voxel_ID_r,
+                              const FreeVec3& ray_sphere_vector, double t, double v,
+                              bool previous_transition_flag) noexcept {
+    assert(false);
+}
+
+// TODO(cgyurgyik): Documentation, implementation.
+[[nodiscard]] AngularHitParameters angularHit(const Ray& ray, const SphericalVoxelGrid& grid,
+        size_t current_voxel_ID_theta, double t) noexcept {
+    assert(false);
+}
+
+// TODO(cgyurgyik): Documentation, implementation.
+[[nodiscard]] AzimuthalHitParameters azimuthalHit(const Ray& ray, const SphericalVoxelGrid& grid,
+        size_t current_phi_ID_theta, double t) noexcept {
+    assert(false);
+}
+
+[[nodiscard]] std::vector<SphericalVoxel>
+sphericalCoordinateVoxelTraversal(const Ray &ray, const SphericalVoxelGrid &grid, double t_begin,
+                                       double t_end, double tol) noexcept {
+    std::vector<SphericalVoxel> voxels;
+    voxels.reserve(grid.numRadialVoxels() + grid.numAngularVoxels() + grid.numAzimuthalVoxels());
+
+    /* INITIALIZATION PHASE */
+
+    // Determine ray location at t_begin.
+    const BoundVec3 point_at_t_begin = ray.point_at_parameter(t_begin);
+    const FreeVec3 ray_sphere_vector = grid.sphereCenter() - point_at_t_begin;
+    double current_r = grid.deltaRadius();
```

Comment:
Should `current_r` be set to `grid.deltaRadius()`?  Maybe I'm misunderstanding, but I assumed that the `deltaRadius` would be `dr` and not the edge of the radius.

Reply:
```
    while (ray_sphere_vector_dot > (current_r * current_r) && current_r < grid.sphereMaxRadius()) {
        current_r += grid.deltaRadius();
    }
```
We are trying to find the radial shell that contains the parameter point of the ray at ```t_begin``` in the next section, so we start ```current_r``` at the innermost radius, grid.deltaRadius() and increment by a factor of ```deltaRadius()``` with each iteration.